dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -136,38 +140,40 @@ class SyncTunnelSender : public TunnelSender
 public:
     using Base = TunnelSender;
     using Base::Base;
-    virtual ~SyncTunnelSender();
-    void startSendThread();
+    ~SyncTunnelSender() override;
+    void startSendThread(PacketWriter * writer);
 
 private:
     friend class tests::TestMPPTunnel;
-    void sendJob();
+    void sendJob(PacketWriter * writer);
     std::shared_ptr<ThreadManager> thread_manager;
 };
 
 /// AsyncTunnelSender is mainly triggered by the Async PacketWriter which handles GRPC request/response in async mode, send one element one time
 class AsyncTunnelSender : public TunnelSender
+    , public GRPCSendQueue<MPPDataPacketPtr>
```

Comment:
please either use private inheritance or use composition.

Reply:
Change to use composition.